### ISSUE TITLE
Loadout problems

### ISSUE THREAD

**Issue by drpepper240** [NoImageAvailable/CombatExtended#385]
Not sure if this is a bug, or everything is working as intended.

Firstly, loadout assingment and especially weapon pickup. I thought it should work similar to vanilla outfit system: after the outfit preset for a certain pawn is assigned or changed, the pawn will try to change its clothes accordingly just after it finishes its current job. After the loadout is assigned, the pawn might pick some stuff like medicine up, but never weapons and seldom any ammo: I have to manually force each pawn to take a weapon of the assigned type. The pawn will then pick up ammo for this weapon according to its loadout automatically. Is this intended behaviour?

Second: when a loadout is changed to an empty loadout or to "Nothing", the pawn won't drop its weapon and inventory contents until I force it to do so in "drafted" mode. If undrafted, the pawn sometimes picks its weapon again just after it was dropped. This probably is a bug, [because](https://github.com/NoImageAvailable/CombatExtended#combat-extended)
> Colonists with a loadout will automatically clear their inventory of surplus items and pick up anything they're missing. Use them to automate picking up ammo, grenades and other consumables.

**Comment by NoImageAvailable** [comment]
It sounds like you have the auto-equip option enabled in the mod settings. If so, try turning it off and let me know if the problem persists.

**Comment by drpepper240** [comment]
Yes, this option was enabled. Turned it off, restarted the game and loaded my savefile, then retested:
- switched from "Nothing" to a firearm+ammo+meds loadout: pawns automatically pick up nearby (<=19 tiles excluding item tile and pawn tile) items, but ignore distant (>19 tiles) items.
- switched from firearm+ammo+meds loadout to "Nothing" for a fully-equipped pawn: the pawn doesn't dequip its gear automatically even if close to a storage zone with max priority. If the pawn is forced to drop its weapon, it does so, doesn't pick it up, but does not drop ammo automatically.
- switched from firearm+ammo+meds loadout to an empty loadout: same as above.

I'm using version 0.17.1.7, I think it's the current master branch.

**Comment by Alicecomma** [comment]
- It seems that the loadout system 1) has a huge preference for items up to 20 tiles away, 2) will assign priority to items up to 80 tiles away as follows: 2.1) assign low stock if there's less than half of the loadout's amount, 2.2) assign low otherwise (e.g high stock or food). Things of "Low" priority have a priority of 3f while any others have a priority of 9.2f.
This means that guns, which are generally of "Low" stock, are assumed to be of much lower priority than ammo, which is generally in low stock.

- When loadouts are _switched_, pawns will try to meet that loadout's requirements. The loadout "Nothing" isn't technically a loadout and it will disable the loadout behaviour altogether and switch over to TakeAndEquip behaviour (basically hoarding weapons), so it won't unequip anything but pick up nearby weapons. This is at least the design.

**Comment by NoImageAvailable** [comment]
Loadout priorities are fixed in https://github.com/NoImageAvailable/CombatExtended/commit/2611050b9a819eaee7ac9e26e63a79586e3af5a0

A wrong operator caused prioritization to be reversed from what they should be.

**Comment by drpepper240** [comment]
Thank you.
I've applied changes for _JobGiver_UpdateLoadout.cs_ from https://github.com/NoImageAvailable/CombatExtended/commit/2611050b9a819eaee7ac9e26e63a79586e3af5a0 to https://github.com/NoImageAvailable/CombatExtended/tree/498e8c4b840788ea5d6fca2ecdc23595ca751fd3 and built it. The problem with pawns not taking/dropping items when loadouts are switched is gone, at least in A17.

> The loadout "Nothing" isn't technically a loadout (null) [and so is an empty loadout] and it will disable the loadout behaviour altogether and switch over to TakeAndEquip behaviour (basically hoarding weapons), so it won't unequip anything but pick up nearby weapons. This is at least the design.

Since this issue is tagged as "enhancement", may I request an enhancement? A way to create a valid loaout which is really empty would be useful. Currently  the closest solution or workaround is to use a loadout with single small and light item.

**Comment by Alicecomma** [comment]
I'm looking into the enhancement:

- If auto-take ammo is enabled AND the loadout is "Nothing", pawns will keep whatever inventory they have (and improve on it without a loadout).
- If auto-take ammo is disabled AND the loadout is "Nothing", pawns will drop stuff in their inventory unless they have a "hold record", and they keep it empty. This is like reverting to Vanilla but with pawns having a tendency to empty their inventory.
- If the loadout is NOT "Nothing", pawns will do what you'd expect from a loadout. Auto-take ammo doesn't have an effect on the pawn anymore.
